rio_tiler/utils.py

```diff
@@ -388,74 +437,84 @@ def _tile_read(
 
-    if not bounds_crs:
-        bounds_crs = dst_crs
-
-    bounds = transform_bounds(bounds_crs, dst_crs, *bounds, densify_pts=21)
+    if bounds_crs:
+        bounds = transform_bounds(bounds_crs, dst_crs, *bounds, densify_pts=21)
 
-    vrt_params = dict(
-        add_alpha=True, crs=dst_crs, resampling=Resampling[resampling_method]
-    )
-
-    vrt_transform, vrt_width, vrt_height = get_vrt_transform(
-        src_dst, bounds, dst_crs=dst_crs
-    )
-
-    out_window = windows.Window(
-        col_off=0, row_off=0, width=vrt_width, height=vrt_height
-    )
-
-    src_bounds = transform_bounds(src_dst.crs, dst_crs, *src_dst.bounds, densify_pts=21)
-    x_overlap = max(0, min(src_bounds[2], bounds[2]) - max(src_bounds[0], bounds[0]))
-    y_overlap = max(0, min(src_bounds[3], bounds[3]) - max(src_bounds[1], bounds[1]))
-    cover_ratio = (x_overlap * y_overlap) / (
-        (bounds[2] - bounds[0]) * (bounds[3] - bounds[1])
-    )
-    if minimum_tile_cover and cover_ratio < minimum_tile_cover:
-        raise TileOutsideBounds(
-            "Dataset covers less than {:.0f}% of tile".format(cover_ratio * 100)
+    if minimum_tile_cover:
+        src_bounds = transform_bounds(
+            src_dst.crs, dst_crs, *src_dst.bounds, densify_pts=21
         )
-
-    if tile_edge_padding > 0 and not _requested_tile_aligned_with_internal_tile(
-        src_dst, bounds, tilesize
-    ):
-        vrt_transform = vrt_transform * Affine.translation(
-            -tile_edge_padding, -tile_edge_padding
+        x_overlap = max(
+            0, min(src_bounds[2], bounds[2]) - max(src_bounds[0], bounds[0])
         )
-        orig_vrt_height = vrt_height
-        orig_vrt_width = vrt_width
-        vrt_height = vrt_height + 2 * tile_edge_padding
-        vrt_width = vrt_width + 2 * tile_edge_padding
-        out_window = windows.Window(
-            col_off=tile_edge_padding,
-            row_off=tile_edge_padding,
-            width=orig_vrt_width,
-            height=orig_vrt_height,
+        y_overlap = max(
+            0, min(src_bounds[3], bounds[3]) - max(src_bounds[1], bounds[1])
         )
+        cover_ratio = (x_overlap * y_overlap) / (
+            (bounds[2] - bounds[0]) * (bounds[3] - bounds[1])
+        )
+        if cover_ratio < minimum_tile_cover:
+            raise TileOutsideBounds(
+                "Dataset covers less than {:.0f}% of tile".format(cover_ratio * 100)
+            )
 
-    vrt_params.update(dict(transform=vrt_transform, width=vrt_width, height=vrt_height))
-
-    indexes = indexes if indexes is not None else src_dst.indexes
-    out_shape = (len(indexes), tilesize, tilesize)
+    overview_level = get_overview_level(src_dst, bounds, tilesize)
+    options = {"overview_level": overview_level} if overview_level is not None else {}
+    with rasterio.open(src_dst.name, **options) as ovr_dst:
+        w, s, e, n = bounds
+        vrt_transform = transform.from_bounds(w, s, e, n, tilesize, tilesize)
```

Comment:
This is the main change.
Instead of relying on WarpedVRT, we force rasterio to use a specific overview level (or the highest resolution)

Note, we cannot pass `overview_level=None` to rasterio because it somehow get converted to `overview_level=0`

Reply:
@vincentsarago the huge slowdown you saw was caused by not using overviews?

Reply:
I'm still not sure if this is going to fix it or not.

>  the huge slowdown you saw was caused by not using overviews?
I don't think rio-tiler **wasn't** fetching the overview.

I'm still investigating, this is one path to follow but not 💯 sure what's going on! I'll try to come up with some repeatable tests case during the week end